lib/active_record/connection_adapters/sqlserver/database_statements.rb

```diff
@@ -19,11 +19,10 @@ def exec_query(sql, name = 'SQL', binds = [], sqlserver_options = {})
           sp_executesql(sql, name, binds)
         end
 
-        def exec_insert(sql, name, binds, _pk = nil, _sequence_name = nil)
-          id_insert = binds_have_identity_column?(binds)
-          id_table  = table_name_from_binds(binds) if id_insert
-          if id_insert && id_table
-            with_identity_insert_enabled(id_table) { exec_query(sql, name, binds) }
+        def exec_insert(sql, name, binds, pk = nil, _sequence_name = nil)
+          id_insert_table_name = query_requires_identity_insert?(sql) if pk
+          if id_insert_table_name
```

Comment:
If you still want to check that id_insert_table_name is not nil you could also do it like
```
if pk && (id_insert_table_name = query_requires_identity_insert?(sql))
  ...
end
```
This would check for pk and if it passes do the assignment and check. Parentheses may not be needed but clearly show the process.